[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91343
* __->__ #91234

**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

fegin committed:
[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

rohan-varma commented on test/distributed/_composable/test_fully_shard.py:
```diff
@@ -30,7 +29,13 @@
+    FSDPInitMode,
+    FSDPTest,
+    TransformerWithSharedParams,
+)
```
nit: can we have formatting changes in separate PR?

I recognize this is tricky and I think it's time to align on formatting convention for FSDP codebase and automate it. cc @awgu

awgu replied:
My plan was to just get everyone on `lintrunner` and `lintrunner f` beginning of next half. I decided since we are cranking PRs with urgency right now, we can just not worry about it. The PR to achieve this look like: https://github.com/pytorch/pytorch/pull/90873 I have re-pushed recently, but the main change is just in the `.lintrunner.toml` file and making sure all relevant files are compliant.

I do think that unifying under `lintrunner` / `lintrunner f` is nice. Sometimes I add changes to a file that create long lines or add imports, and I want to just auto-format. However, without an agreed-upon auto-formatter, this becomes a problem and actually complicates the workflow.

fegin replied:
Will rebase this PR on top of https://github.com/pytorch/pytorch/pull/91255.

rohan-varma commented on test/distributed/_composable/test_fully_shard.py:
```diff
@@ -589,12 +599,57 @@ def _check_model_parity(self, m1: nn.Module, m2: nn.Module):
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def _test_optim_state_dict_save_load(self):
```
Might be better to just have `test` instead of disabling with prefix, and adding skip decorator mentioning reason it is disabled and

rohan-varma commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -87,10 +87,7 @@ def fully_shard(
-        ):
-            _insert_module_state(submodule, state)
+    # Insert all comm_modules to the module to state mapping.
+    for submodule in state._fully_sharded_module_to_handles.keys():
```
Is this change equivalent to the former code? If not, is there a reasoning we're changing the inserted states?

fegin replied:
This is not equivalent to the former code. The reason behind the change is to only map the modules that actually have the handles -- the local root modules.

rohan-varma commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1280,8 +1275,20 @@ def _optim_state_dict(
+    mapping between parameters and parameter IDs. Using ``optim_input`` is being
+    deprecated.
+
+    If the optimizer is a ``NamedOptimizer``, the optimizer state_dict does not
```
what if optim_input is provided but also it is a NamedOptimizer, will that create an issue?

fegin replied:
Yes, it will fail. Will add a error handling for this.

rohan-varma commented on test/distributed/_composable/test_fully_shard.py:
```diff
@@ -589,12 +599,57 @@ def _check_model_parity(self, m1: nn.Module, m2: nn.Module):
+            composable_optim_state_dict["param_groups"],
+        ):
+            for key, value in group1.items():
+                self.assertEqual(value, group2[key])
```
Is it worth adding tests for:

- non root FSDP
- DDP / replicate root
- nested FSDP + non root?

fegin replied:
Add an extra test for non root FSDP. Will add more tests after fixing the `all_gather_object` issue that prevent us from running tests on CI.

rohan-varma reviewed (approved):
LGTM! Super exciting stuff :D

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][8/N] Enable fully_shard optim state_dict save and load"


**What does this PR do?**
This PR refactor `_optim_utils.py` to use `_FSDPState` instead of `FullyShardedDataParallel` class. This change enables the support of optim state_dict for `fully_shard`.

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge
